---
title: KeyPathwayMineR - R Package for the web and standalone version of KeyPathwayMiner
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{KeyPathwayMineR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r, include = FALSE}
# General markdown options
# Collapses output and comments it to 
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
```{r quick_load_data, include = FALSE}
library(KeyPathwayMineR)
colon_expression_up <- as.data.frame.matrix(read.delim(
  system.file(package = "KeyPathwayMineR",
              "extdata/datasets",
              "colon-gene-expression-UP-p0.05.txt"),
  header = FALSE))

huntington_disease_up <- as.data.frame.matrix(read.delim(
  system.file(package = "KeyPathwayMineR",
              "extdata/datasets",
              "huntington-gene-expression-UP.txt"),
  header = TRUE))

huntington_disease_up <- to_indicator_matrix(numerical_matrix = huntington_disease_up,
                                             operator = "<", threshold = 0.005)

sample_network = system.file("extdata", "sampleNetwork.sif", package="KeyPathwayMineR")
# Path to colon methylation matrix
colon_methylation <- system.file(package = "KeyPathwayMineR",
                                            "extdata/datasets",
                                            "colon-methylation-matrix-p0.05.txt")
```
## I. Introduction
### KeyPathwayMiner
Given a **biological network** and a **set of case-control studies**, KeyPathwayMiner(KPM) efficiently extracts all             **maximal connected sub-networks**. These sub-networks contain the genes that are **mainly** dysregulated, e.g.,           differentially expressed, in **most** cases studied:

   * The exact quantities for “mainly” and “most” are modeled with two easy-to-interpret parameters **(K, L)** that allows      the user to control the number of outliers (not dysregulated genes/cases) in the solutions.
     
Two different approaches for extracting subnetworks that are enriched for active/deregulated genes have been               implemented:

  * **INES:**  Extract all maximal sub-networks containing nodes with no more than L inactive cases (0's) besides of K                    exceptions.
      
  * **GLONE:** Extracts maximal sub-networks where the total sum of **not**-active/diff. exp. cases is at most L. 
      
For more information please visit the [KPM website](https://exbio.wzw.tum.de/keypathwayminer/).

### KeyPathwayMineR
The following package provides an easy way for using KPM and visualize the extracted subnetworks via R.
The user can decide whether he wants to call KPM on his computer (via standalone) or with the Web API. 
With the `kpm_options()` object the arguments for the execution of the program can be easily processed and viewed.


## II. Setup KeyPathwayMineR
KeyPathwayMineR can be currently installed from github by using the devtools package
```{r download_from_github_repo, eval = FALSE}
library(devtools)
install_github(repo = "konstantinosM/KeyPathwayMineR", build_vignettes = TRUE)
```
If installation was successful the package can be loaded via
```{r load_KeyPathwayMineR}
library(KeyPathwayMineR)
```
**Prerequisites:**

 * When loading the package a script will test if your R is using java jvm 1.8.
 * And if the standalone jar was correcly attached to your classpath.
 
If **both** requirements are met you are ready to go.

## III. Getting started
In the following section, we will take a look at how to run KeyPathwayMineR **remotely** and **locally**.  We will play around with the options manager **kpm_options()** which contains all parameters needed for running KPM. Finally, we will prepare different input data and go through multiple examples of how to run KeyPathwayMineR with different parameters. 

### Options
The options manager provides the user with a variety of arguments he can customize for his execution. To get a documentation and overview of all parameters available you can run the following commands:
```{r options_manager}
# For the documentation
?kpm_options()

# For a  general overview of all parameters use without head
head(kpm_options())
```
You can change the options by choosing the option you want and assigning it to a new value like that:
```{r change_option}
# Initial value
kpm_options()$algorithm

# Change algorithm to ACO
kpm_options(algorithm = "ACO")

# After changing the value
kpm_options()$algorithm

# You can also change multiple parameters by seperating them with a comma
kpm_options(algorithm = "Optimal", execution = "Local")
```
Reset your options for a new run like that:
```{r reset_option}
# Initial value
reset_options()
```

### Input data
Multiple input files can be provided to use in KeyPathwayMineR, but the most important ones are the:

 * **matrix file/s** from the expression studies. Either numeric (e.g. p-values or fold changes) or as indicator matrix.
 * **graph file** containing all the interactions of the protein interaction network. 
 
The other files are only relevant for the local execution. For detailed information on all the files and their exact input format use `vignette("input_files_format")`.

#### Dataset file/s
Lets get started with the dataset files (matrix files):
```{r datasets_files}
# Read colon-gene-expression-up-05 indicator_matrix as data.frame. 
colon_expression_up <- as.data.frame.matrix(read.delim(
                                       system.file(package = "KeyPathwayMineR",
                                                   "extdata/datasets",
                                                   "colon-gene-expression-UP-p0.05.txt"),
                                       header = FALSE))

# Preview colon expression indicator matrix
colon_expression_up[1:5,1:15]

# Read huntington_disease_up as data.frame
huntington_disease_up <- as.data.frame.matrix(read.delim(
                                        system.file(package = "KeyPathwayMineR",
                                                    "extdata/datasets",
                                                    "huntington-gene-expression-UP.txt"),
                                        header = TRUE))

# Preview huntigton disease up numerical matrix
huntington_disease_up[1:5,1:6]
```
The entries of the huntington_disease_up dataset are p-values. To create an indicator matrix we will use the to_indicator_matrix() function as follows: 
```{r to_indicator_matrix}
# Convert numerical to indicator matrix
huntington_disease_up <- to_indicator_matrix(numerical_matrix = huntington_disease_up,
                                             operator = "<", threshold = 0.005)

# Huntigton disease up
huntington_disease_up[1:5,1:9]
```
You can also just provide the path to an indicator matrix file.  Assuming the matrix is provided in the right format KeyPathwayMineR will automatically detect the type of input and processes it according to the execution type (local or remote):
```{r path_to_indicator_matrix}
# Path to colon methylation matrix
colon_methylation <- system.file(package = "KeyPathwayMineR",
                                            "extdata/datasets",
                                            "colon-methylation-matrix-p0.05.txt")
```

#### Graph file
In the next step, we will go through how to prepare the network files for the execution.

Firstly the graph_file can be provided as a path:
```{r graph_file_path}
sample_network = system.file("extdata", "sampleNetwork.sif", package="KeyPathwayMineR")
```
For the remote execution, the user can as well select a graph_id from the web service.
```{r get_networks}
# Get overview of available networks
get_networks()
```
Now the user just needs to save the graph_id from the graph he wants to use:
```{r graph_id}
# We will choose the Krogan Sars-COV2 + IID human host interactome whith graph_id = 13
kpm_options(graph_id = 13)
```
### Execution 
Once all the input files are prepared we are ready to start de novo pathway enrichment analysis.

#### Remote
Lets take a look which URL KeyPathwayMineR will use for the HTPP requests.
```{r intialize}
# KeyPathWayMiner URL:
kpm_options()$url
```
Start with a simple run on one dataset with fixed parameters for K and L using INES in a blocking call(async = FALSE). Blocking meaning the user will have to wait until the execution is complete. For this run we will use the graph_id 10 which is the "I2D Homo_sapiens entrez" graph.
```{r remote_execution_example1, eval = FALSE}
# Configure options for the run
kpm_options(execution = "Remote",
            async = FALSE,
            strategy = "INES",
            remove_bens = TRUE,
            algorithm = "Greedy",
            l_min = 8,
            k_min = 1,
            graph_id = 10)

# Start run with huntington_disease_up dataset
run <- kpm(indicator_matrices = huntington_disease_up)

# Get your results from
print(paste("Browse to", run$resultUrl, "to see the results"))

# Reset setting for next run
settings::reset(kpm_options)
```

----

In the next example we will perfom a ranged run in a blocking call using INES. 
By setting `use_range_k = TRUE` and `use_range_l = TRUE` you 
specify that the parameters L and K should be ranged.

**Comment:** `l_min = 4`, `l_max = 8` and `l_step = 2` would mean that the KPM will run with L = 4, L = 6 and L = 8.

When `use_range` is set to **FALSE** only l_min is relevat. The same applies to K.
```{r remote_execution_example2, eval = FALSE}
# Configure options for the run
kpm_options(execution = "Remote",
            async = FALSE,
            strategy = "INES",
            remove_bens = TRUE,
            algorithm = "Greedy",
            use_range_l = TRUE, l_min = 4, l_step = 2, l_max = 8,
            use_range_k = TRUE, k_min = 0, k_step = 1, k_max = 2,
            graph_id = 13)

# Start run with colon_expression_up dataset
run <- kpm(indicator_matrices = colon_expression_up)

# Get your results from
print(paste("Browse to", run$resultUrl, "to see the results"))

# Reset setting for next run
settings::reset(kpm_options)
```

----

In the last example for the remote execution, we will take a look at how to combine and use multiple datasets. We will start a run with INES, K = 1 and L = 10% for the two datasets huntington_disease_up and colon_expression_up. This time the job will be submitted asynchronously. This means the user will not have to wait for the job to complete. He can check on a later time if the job is complete.
```{r remote_execution_example3a, eval = FALSE}
kpm_options(execution = "Remote",
            async = TRUE,
            strategy = "INES",
            remove_bens = TRUE,
            algorithm = "Greedy",
            l_same_percentage = TRUE, same_percentage = 10,
            k_min = 1, 
            graph_id = 13,
            link_type = "OR")
```  
You should have noticed that the link_type was set to OR. This means the two datasets will be logically connected with the "OR" operator.
```{r remote_execution_example3b, eval = FALSE}
# Add the two datasets to a list
indicator_matrices <- list(colon_expression_up, huntington_disease_up)

# Run kpm 
run <- kpm(indicator_matrices = indicator_matrices)

# Extract job ID (called quest id)
quest_id <- run$questID

# Open the result page where you can  monitor the progress of both tasks
print(quest_progress_url(session_id = kpm_options()$session_id))

#Get status of the Job and if complete download the resuls
if(get_status(quest_id = quest_id)$completed){
  results_example_3 <- get_results(quest_id = quest_id)
} 

# Reset settings
settings::reset(kpm_options)
```

**Important note:** The **web** service does not allow individual fixed (L case-exceptions) parameters to be set for each matrix at the moment. Robustness analysis is also not supported via Remote execution. If you want to use one of those two things then please use the **Local** execution type.

#### Local
For users who want to conduct more extensive analysis, the local execution is the way to go. Not only is the user able to specify different case exceptions parameters (L) for every dataset but he can also analyze the robustness of the provided network by choosing one of the four perturbation techniques.

For the users who want to go even more into detail, we provide a properties file where the user can specify the default parameters for the local run. To access and edit the properties file use `edit(file = system.file(package = "KeyPathwayMineR", "kpm.properties"))`. Keep in mind that options set with kpm_options() have higher priority than the options in the kpm.properites file.

In the first example, we will perform an unranged run using INES. In addition to that we will analyse the robustness of the network with the perturbation technique "nodeswap" and the following parameters:
```{r With perturbation, eval = FALSE}
# Use perturbation
kpm_options(execution = "Local", strategy = "INES", algorithm ="Greedy",l_min=5,
            k_min = 2, perturbation_technique ="nodeswap",
            with_perturbation = TRUE,
            perturbation_start = 10,
            perturbation_step = 10,
            perturbation_max = 20,
            graphs_per_step = 1,
            resource_folder = system.file(package = "KeyPathwayMineR", "extdata/"),
            properties_file = system.file(package = "KeyPathwayMineR",  "kpm.properties"))

# Execute remote run by using a custom graph_file
kpm(graph_file = sample_network, indicator_matrices  = huntington_disease_up)

settings::reset(kpm_options)
```

----

In this example, an unranged GLONE run will be demonstrated for multiple datasets. Since we are using two datasets we will have to define two l parameters. This is possible by providing a vector as input for l_min:
```{r Combine multiple datasets, eval = FALSE}
indicator_matrices = list(huntington_disease_up, colon_expression_up)
# Combine multiple datasets:
kpm_options(execution = "Local", 
            strategy = "GLONE",
            algorithm = "Greedy",
            l_min = c(5, 2),
            resource_folder = system.file(package = "KeyPathwayMineR", "extdata/"),
            properties_file = system.file(package = "KeyPathwayMineR",  "kpm.properties"))
# Run kpm
kpm(graph_file = sample_network, indicator_matrices  = indicator_matrices)

# Reset settings
settings::reset(kpm_options)
```

---

In the last example, we will perform a ranged (batch) rung using INES. It is the same idea as in the example before, only this time we also include l_step and l_max:
```{r Batch run, eval = FALSE}
indicator_matrices = list(colon_methylation, colon_expression_up)

# Use ranged values with batch
kpm_options(execution = "Local",
            strategy = "INES",
            algorithm ="Greedy",
            l_min = c(1,3),
            l_step = c(2,4),
            l_max = c(3,6),
            k_min = 1,
            k_step = 2,
            k_max = 3,
            use_range_k = TRUE,
            use_range_l = TRUE,
            resource_folder = system.file(package = "KeyPathwayMineR", "extdata/"),
            properties_file = system.file(package = "KeyPathwayMineR",  "kpm.properties"))

# Run kpm
kpm(graph_file = sample_network, indicator_matrices  = indicator_matrices)

# Reset settings
settings::reset(kpm_options)
```
For the first datasets we are using `l_min = 1, l_step = 2, l_max = 3` and for the second dataset
`l_min = 3, l_step = 4 and l_max = 6`.

**Note:** The results of the local run will be saved in the results folder on you working directory. If you wish to save you results in another place, change the results_folder option in the properties file.

## IV. Conclusion
Our goal is to simplify the user's work process when using KeyPathwayMiner. Users can perform their analysis directly in R without having to worry about how to access Jar files or set up HTTP requests. 
At the same time, R provides an efficient way to prepare, handle and further process data. This is for example very practical for the preparation of indicator matrices and other input data. Finally, the user can choose whether he wants to use the remote or the local version depending on the performance of his computer and his requirements.

